### PR TITLE
Openai agent json malformed 11921

### LINKED ISSUE #11921: [Bug]: Agent errors when json function argument is malformed
### Bug Description

We use the OpenAIAgent together with some custom tools.
We noticed, that even gpt4-turbo has some issues with creating a valid json in the function calls when the prompt gets big.
For example, the tool has an input named "query" which should be a string (in our concrete case, python code to be executed). So the function arguments string should look like '{"query": "some long string"}', but often is malformed - e.g. 'query = """some long multi-line string"""'.
Currently, this results in an error in the agent, as in agent.openai.step.call_function, the argument_str is parsed via json.loads, without error handling - which fails in those circumstances and a JSONDecodeError is thrown.

### Version

0.10.19

### Steps to Reproduce
```
from llama_index.agent.openai import OpenAIAgent
from dotenv import load_dotenv
from llama_index.llms.openai import OpenAI
from llama_index.core.tools.function_tool import FunctionTool
from pydantic import BaseModel, Field

load_dotenv()


class PythonInput(BaseModel):
    query: str = Field(
        description="python script or command WIHTOUT COMMENTS which will be evaluated by the eval command.\nNever start variable names with numbers!'"
    )


def python_tool(query: str) -> str:
    # acutally, a fake python tool
    print("Python tool called with query:")
    print(query)
    print("______")
    return query


tool = FunctionTool.from_defaults(
    name="python-tool",
    fn=python_tool,
    description="A Python shell. Use this to execute python commands.\nNever start variable names with numbers!\n",
    fn_schema=PythonInput,
)

long_system_message = """You are working with a python shell that has a pandas DataFrame.
The name of the dataframe is `df`.
You have pandas and numpy available as pd and np.This is the description of df:

This DataFrame contains balance sheet data of the company in question for several years.
Each row contains the balance sheet data of a year.
The column 'Year' is very important. It is sorted DESC and contains the year for which the report is valid.
The column 'Currency' is also important, it specifies the currency in which the numbers are reported in.

Some rules to follow:

1. Group calculations per Year and include the Year column in your answer.
2. If you are asked how a metric developed, respond with the absolute values of the last couple of years, not the percentage changes.
3. Return information from the context in the form of a markdown table.
4. Include the Currency column if possible.
5. Don't use dropna on df, you could lose important information.
6. If calculating diffs: use .diff(-1) to calculate differences to previous year. And don't include the Currency column for the diff, only in the assign part. Good example:
df[['Year', 'Total_Assets', 'Cash_Bank_Deposits', 'AS30', 'Goodwill', 'AS32', 'AS40']].diff(-1).assign(Year=df['Year'], Currency=df['Currency'])

This is the description of the relevant columns:
Column 'Trade_Receivables': Trade Receivables 
Column 'Intragroup_Receivables': Intragroup Receivables
Column 'Other_Receivables': Other Receivables
Column 'Subtotal_Inventory': Subtotal Inventory
Column 'Total_Assets': Total Assets 
Column 'Currency': Currency in which all figures for this balance sheet are reported

This is the result of `print(df.head())`:
   Other_Receivables Currency  Year  Total_Assets  Intragroup_Receivables  Subtotal_Inventory  Trade_Receivables
0       2.025012e+09      EUR  2022  1.551012e+10                       0        7.689101e+08       6.813689e+08
1       3.801111e+08      EUR  2019  1.647862e+10                       0        6.358830e+08       7.145821e+08

Use the tool to answer the questions posed to you."""

# Here is a question about the dataframe, that needs multiple lines of code to answer:
question = "For all of the following lines, calculate its share of total assets: Subtotal Liquid Assets, Trade Receivables, Intragroup Receivables, Other Receivables, Subtotal Inventory, Subtotal Tangible Fixed Assets, Subtotal Intangible Fixed Assets, Subtotal Financial Fixed Assets"
# in our production use case, we tried our best to not confuse the llm with the python_tool docstring or the system message, but the llm still got confused.
for i in range(10):
    llm = OpenAI(
        model="gpt-4-1106-preview", temperature=0.0000000001, additional_kwargs={"seed": i, "top_p": 0.0000000001}
    )
    agent = OpenAIAgent.from_tools(tools=[tool], system_prompt=long_system_message, llm=llm, verbose=True)
    print(f"Seed: {i}")
    response = agent.chat(question)
    print(response)

```

### Relevant Logs/Tracbacks

ERROR:root:Error while running agent.chat("For all of the following lines, calculate its share of total assets: Subtotal Liquid Assets, Trade Receivables, Intragroup Receivables, Other Receivables, Subtotal Inventory, Subtotal Tangible Fixed Assets, Subtotal Intangible Fixed Assets, Subtotal Financial Fixed Assets")
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/core/callbacks/utils.py", line 41, in wrapper
    return func(self, *args, **kwargs)
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/core/agent/runner/base.py", line 575, in chat
    chat_response = self._chat(
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/core/agent/runner/base.py", line 520, in _chat
    cur_step_output = self._run_step(
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/core/agent/runner/base.py", line 372, in _run_step
    cur_step_output = self.agent_worker.run_step(step, task, **kwargs)
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/core/callbacks/utils.py", line 41, in wrapper
    return func(self, *args, **kwargs)
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/agent/openai/step.py", line 572, in run_step
    return self._run_step(
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/agent/openai/step.py", line 469, in _run_step
    self._call_function(
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/agent/openai/step.py", line 361, in _call_function
    function_message, tool_output = call_function(
  File "/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/llama_index/agent/openai/step.py", line 99, in call_function
    argument_dict = json.loads(arguments_str)
  File "/usr/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

```shell
Added user message to memory: For all of the following lines, calculate its share of total assets: Subtotal Liquid Assets, Trade Receivables, Intragroup Receivables, Other Receivables, Subtotal Inventory, Subtotal Tangible Fixed Assets, Subtotal Intangible Fixed Assets, Subtotal Financial Fixed Assets
=== Calling Function ===
Calling function: python-tool with args: query = """
df.assign(
    Subtotal_Liquid_Assets_Share=df['Cash_Bank_Deposits'] / df['Total_Assets'],
    Trade_Receivables_Share=df['Trade_Receivables'] / df['Total_Assets'],
    Intragroup_Receivables_Share=df['Intragroup_Receivables'] / df['Total_Assets'],
    Other_Receivables_Share=df['Other_Receivables'] / df['Total_Assets'],
    Subtotal_Inventory_Share=df['Subtotal_Inventory'] / df['Total_Assets'],
    Subtotal_Tangible_Fixed_Assets_Share=df['AS40'] / df['Total_Assets'],
    Subtotal_Intangible_Fixed_Assets_Share=df['Goodwill'] + df['AS32'] / df['Total_Assets'],
    Subtotal_Financial_Fixed_Assets_Share=df['AS30'] / df['Total_Assets']
)[['Year', 'Currency', 'Subtotal_Liquid_Assets_Share', 'Trade_Receivables_Share', 'Intragroup_Receivables_Share', 'Other_Receivables_Share', 'Subtotal_Inventory_Share', 'Subtotal_Tangible_Fixed_Assets_Share', 'Subtotal_Intangible_Fixed_Assets_Share', 'Subtotal_Financial_Fixed_Assets_Share']]
"""
```

### PR DESCRIPTION
**Pull request by JoelGotsch**
# Description

The OpenAIAgent no longer crashes when the function argument is not valid json. Furthermore, a tool_call_parser can be passed when creating the OpenAIAgent that allows the user to handle/ recover from malformed tool calls. For an example, see docs/examples/agent/openai_agent_tool_call_parser.ipynb

Fixes  #11921

## Version Bump?
- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] Added unit/integration tests and executed them locally

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/agent/llama-index-agent-openai/llama_index/agent/openai/step.py [hunk 1]
```diff
@@ -446,6 +549,12 @@ def _run_step(
         openai_tools = [tool.metadata.to_openai_tool() for tool in tools]
 
         llm_chat_kwargs = self._get_llm_chat_kwargs(task, openai_tools, tool_choice)
+        for tool in tools:  # self.agent_worker._get_tools("test"):
```
Comment (logan-markewich) [inline llama-index-integrations/agent/llama-index-agent-openai/llama_index/agent/openai/step.py:552]: Are these debug prints? Can we remove them?

File: docs/examples/agent/openai_agent_tool_call_parser.ipynb
Comment (logan-markewich) [inline docs/examples/agent/openai_agent_tool_call_parser.ipynb:1]: Do you mind rerunning this? I think there are some debug prints (i.e. the `NOT GIVEN` stuff, and others), makes it a little confusing to read the outputs

File: llama-index-integrations/agent/llama-index-agent-openai/llama_index/agent/openai/step.py (hunk 1 above)
Comment (JoelGotsch) [reply llama-index-integrations/agent/llama-index-agent-openai/llama_index/agent/openai/step.py:552]: Indeed, removing them in the next commit

File: docs/examples/agent/openai_agent_tool_call_parser.ipynb
Comment (JoelGotsch) [reply docs/examples/agent/openai_agent_tool_call_parser.ipynb:1]: No problem at all

**Comment by JoelGotsch** [comment]
@logan-markewich just merged with current main and pushed the changes.

**Comment by nerdai** [comment]
@JoelGotsch could you kindly pull latest from main and resolve conflicts?

**Comment by JoelGotsch** [comment]
@nerdai done

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**